[lexical][lexical-selection][lexical-utils] Refactor: Consolidate ancestor lookup via findMatchingParent

cnaples79 opened the pull request:
<!-- 
Title format should be:
[Affected Packages] PR Type: title

Example:
[lexical-playground][lexical-link] Feature: Add more emojis 

Choose from the following PR Types:
Breaking change / Refactor / Feature / Bug Fix / Documentation Update / Chore
-->

## Description
<!-- 
- What is the current behavior that you are modifying? 
- What are the behavior or changes that are being added by this PR?
-->

- Current behavior: `$findMatchingParent` and `$getAncestor` logic lived in multiple places (`@lexical/utils`, `@lexical/selection`), leading to duplicated implementations and layering violations. This duplication also surfaced as duplicate symbol/implementation errors in CI when both versions were in play.

- Changes in this PR:
  - Move a single, typed implementation of `$findMatchingParent` into core (`packages/lexical/src/LexicalUtils.ts`) with overloads (type‑predicate support).
  - Export `$findMatchingParent` from core (`packages/lexical/src/index.ts`).
  - `@lexical/utils`: re‑export `$findMatchingParent` from `lexical` and remove the duplicate implementation.
  - `@lexical/selection`: provide a local exported `$getAncestor` wrapper that delegates to core `$findMatchingParent` (import from `lexical`), avoiding a utils dependency and removing local duplication.
  - Remove stray PR‑only comment in `@lexical/link` (no functional change).

No runtime behavior changes are intended; this is a consolidation/refactor to respect package layering and eliminate duplicate code.

Closes #5311

## Test plan

### Before

- Multiple implementations of ancestor lookup existed across packages.
- CI surfaced TypeScript/esbuild errors like duplicate symbol/implementation for `$findMatchingParent` when both versions compiled.

### After

- A single source of truth for ancestor lookup in core, with utils re‑export and selection delegating wrapper.
- Packages compile against the unified helper; behavior remains the same and is covered by existing unit/e2e tests (e.g., selection, links, lists, tables).

*No visual changes.*

etrepum commented on packages/lexical-selection/src/range-selection.ts:
```diff
@@ -36,6 +36,7 @@ import {
   INTERNAL_$isBlock,
 } from 'lexical';
 import invariant from 'shared/invariant';
+import { $findMatchingParent } from '@lexical/utils';
```
@lexical/selection can't depend on @lexical/utils

cnaples79 commented:
@etrepum I see! Thanks for the feedback, I'll update the fix and PR.

cnaples79 commented:
Thanks for the pointer on package layering, @etrepum. I’ve updated the PR to avoid the circular dependency by moving  into core (), exporting it via , and re-exporting from .\n\nChanges: \n- Add  to  with the same typed overloads and docs.\n- Export from .\n- Update  to import from  (not ).\n- Re-export  from  and remove the duplicate local implementation.\n\nThis keeps  free of a dependency on , removes the duplication, and aligns with the guidance from #5311. Happy to iterate further if you'd prefer a different placement or naming.

cnaples79 commented:
Pushed a follow-up to fix the Vercel failure: removed the duplicate  const and added proper overloads to the existing function in  (single implementation). This should unblock the playground build. If there’s a preferred location or naming for the overloads/docs, happy to tweak.

cnaples79 commented:
Found and fixed the Vercel build failure! 🎉

**Issue**: There were duplicate function declarations causing TypeScript compilation errors:
- Overload declarations on lines 981-988 ✅
- Duplicate `export function` declaration on line 1789 ❌

**Fix**: Removed the `export` keyword from line 1789. The function is already properly exported via the overload declarations above.

The TypeScript pattern should be:
1. Export overload declarations 
2. Single implementation function (without export keyword)

This should resolve the build failure. The playground build was already working, so this was specifically affecting the main lexical package build.

etrepum commented on packages/lexical-link/src/index.ts:
```diff
@@ -701,16 +701,7 @@ export function $toggleLink(
-  }
-  return predicate(parent) ? parent : null;
-}
+// Removed duplicate $getAncestor in favor of shared $findMatchingParent from @lexical/utils
```
No need for this comment, only really makes sense in the context of the PR, not the remaining code
```suggestion
```

etrepum commented on packages/lexical-selection/src/range-selection.ts:
```diff
@@ -646,9 +647,6 @@ export function $getAncestor<NodeType extends LexicalNode = LexicalNode>(
-  }
-  return predicate(parent) ? parent : null;
+  // Delegate to shared implementation to avoid duplication.
+  return $findMatchingParent(node, predicate) as NodeType | null;
```
This doesn't remove `$getAncestor` or fix the type of `$findMatchingParent` per https://github.com/facebook/lexical/issues/5311#issuecomment-2512076499

etrepum commented on packages/lexical-utils/src/index.ts:
```diff
@@ -415,31 +415,7 @@ export type DOMNodeToLexicalConversionMap = Record<
-
-  return null;
-};
+export { $findMatchingParent } from 'lexical';
```
This doesn't work, it's used in this module so you need to import it then separately export it.

Are you running this code to test it? Pretty sure `npm run ci-check` would've caught this pretty quickly

etrepum commented on packages/lexical/src/LexicalUtils.ts:
```diff
@@ -1764,15 +1786,15 @@ export function $splitNode(
   while (curr !== $getRoot() && curr != null) {
     if (findFn(curr)) {
-      return curr;
+      return curr as LexicalNode; // T is inferred via overload when appropriate
```
This cast seems redundant, and all casts are potentially unsafe so you should avoid them unless absolutely necessary

etrepum commented on packages/lexical/src/LexicalUtils.ts:
```diff
@@ -2141,3 +2124,37 @@ export function $create<T extends LexicalNode>(klass: Klass<T>): T {
+): LexicalNode | null => {
+  let curr: ElementNode | LexicalNode | null = startingNode;
+
+  while (curr !== $getRoot() && curr != null) {
```
I know that this is just code moved from elsewhere but this is more correct because getLatest isn't called on startingNode first.

```suggestion
  while (!$isRootNode(curr) && curr != null) {
```

etrepum commented on packages/lexical/src/LexicalUtils.ts:
```diff
@@ -2141,3 +2124,37 @@ export function $create<T extends LexicalNode>(klass: Klass<T>): T {
+ * passed as an argument. eg. if( (node) => node.__type === 'div') ) return true; otherwise return false
+ * @param startingNode - The node where the search starts.
+ * @param findFn - A testing function that returns true if the current node satisfies the testing parameters.
+ * @returns A parent node that matches the findFn parameters, or null if one wasn't found.
```
This just makes the description more precise, its treatment of the startingNode and RootNode are not covered by the existing documentation.

```suggestion
 * @returns `startingNode` or one of its ancestors that matches the `findFn` predicate and is not the `RootNode`, or `null` if no match was found.
```

etrepum reviewed (approved):
Made some suggestions for improvements to the existing code. Audited our existing usage of `$getAncestor` (not publicly exported, and does not depend on the implementation differences with `$findMatchingParent`, so it's fine to remove).

etrepum commented:
This looks good, other than the PR title and description. Can you change these to match the repository's pull request template? It's important for how we put together the release notes.

https://raw.githubusercontent.com/facebook/lexical/refs/heads/main/.github/pull_request_template.md
